# CLASSIFICATION IS DONE ON CATEGORICAL DATA
# # Logistic Regression (LR)
# # Linear Discriminant Ananlysis (LDA)
# # K Nearest Neighbour (KNN)

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns

In [2]:
# Loading Final_data data from previous model 

data = pd.read_csv('C:/Users/MANISH KUMAR/Desktop/Y Finance/ML project/Final_data.csv')

# Logistic Regression with Single predictor

In [3]:
data.head()

College  City  Previous CTC  Previous job change  Graduation Marks  \
0        2     0       55523.0                    3                66   
1        2     0       57081.0                    1                84   
2        3     0       60347.0                    2                52   
3        3     0       49010.0                    2                81   
4        1     0       57879.0                    4                74   

   EXP (Month)       CTC  Role_Manager  
0           19  71406.58             1  
1           18  68005.87             0  
2           28  76764.02             0  
3           33  82092.39             0  
4           32  73878.10             0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1589 entries, 0 to 1588
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   College              1589 non-null   int64  
 1   City                 1589 non-null   int64  
 2   Previous CTC         1589 non-null   float64
 3   Previous job change  1589 non-null   int64  
 4   Graduation Marks     1589 non-null   int64  
 5   EXP (Month)          1589 non-null   int64  
 6   CTC                  1589 non-null   float64
 7   Role_Manager         1589 non-null   int64  
dtypes: float64(2), int64(6)
memory usage: 99.4 KB


In [5]:
data.describe()

College         City  Previous CTC  Previous job change  \
count  1589.000000  1589.000000   1589.000000          1589.000000   
mean      1.975456     0.514160  55518.453744             2.528634   
std       0.838330     0.499957   6655.218445             1.123918   
min       1.000000     0.000000  36990.000000             1.000000   
25%       1.000000     0.000000  50518.000000             2.000000   
50%       2.000000     1.000000  55291.000000             3.000000   
75%       3.000000     1.000000  60109.000000             4.000000   
max       3.000000     1.000000  77911.000000             4.000000   

       Graduation Marks  EXP (Month)            CTC  Role_Manager  
count       1589.000000  1589.000000    1589.000000   1589.000000  
mean          59.855255    39.044682   75353.278798      0.206419  
std           14.935139    14.108875   12587.288237      0.404862  
min           35.000000    18.000000   53020.320000      0.000000  
25%           46.000000    26.000000   66902.350000      0.000000  
50%           60.000000    39.000000   73028.670000      0.000000  
75%           73.000000    51.000000   80588.670000      0.000000  
max           85.000000    64.000000  123416.990000      1.000000

# Logistic Regression with single predictor

In [6]:
from sklearn.linear_model import LogisticRegression

In [7]:
X = data[['CTC']]
y = data['Role_Manager']

In [8]:
clf_lrs = LogisticRegression()

In [9]:
clf_lrs.fit(X,y)

LogisticRegression()

In [10]:
clf_lrs.coef_

array([[-1.44177105e-05]])

In [11]:
clf_lrs.intercept_

array([-4.02856353e-10])

# Using stats method

In [12]:
import statsmodels.api as sn
import statsmodels.discrete.discrete_model as sm

In [13]:
X_cons = sn.add_constant(X)

In [14]:
logit = sm.Logit(y,X_cons).fit()

Optimization terminated successfully.
         Current function value: 0.315316
         Iterations 7


In [15]:
X_cons.head()

const       CTC
0    1.0  71406.58
1    1.0  68005.87
2    1.0  76764.02
3    1.0  82092.39
4    1.0  73878.10

In [16]:
logit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:           Role_Manager   No. Observations:                 1589
Model:                          Logit   Df Residuals:                     1587
Method:                           MLE   Df Model:                            1
Date:                Sat, 27 Jan 2024   Pseudo R-squ.:                  0.3807
Time:                        21:14:55   Log-Likelihood:                -501.04
converged:                       True   LL-Null:                       -809.08
Covariance Type:            nonrobust   LLR p-value:                5.327e-136
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -13.2024      0.676    -19.527      0.000     -14.528     -11.877
CTC            0.0001    8.3e-06     17.987      0.000       0.000       0.000
==============================================================================
"""

# Logistic Regression with Multiple predictor

In [17]:
data.head()

College  City  Previous CTC  Previous job change  Graduation Marks  \
0        2     0       55523.0                    3                66   
1        2     0       57081.0                    1                84   
2        3     0       60347.0                    2                52   
3        3     0       49010.0                    2                81   
4        1     0       57879.0                    4                74   

   EXP (Month)       CTC  Role_Manager  
0           19  71406.58             1  
1           18  68005.87             0  
2           28  76764.02             0  
3           33  82092.39             0  
4           32  73878.10             0

In [18]:
# Classification is done categorical data

X = data.loc[:, data.columns != 'Role_Manager']
y = data['Role_Manager']

In [19]:
clf_lrs = LogisticRegression()

In [20]:
clf_lrs.fit(X,y)

LogisticRegression()

In [21]:
clf_lrs.intercept_

array([-0.00055792])

In [22]:
 clf_lrs.coef_  

array([[-0.00087482, -0.00063605, -0.00021655, -0.00192263, -0.03305709,
        -0.05451868,  0.00018345]])

# Using stats method

In [23]:
import statsmodels.discrete.discrete_model as sm

In [24]:
X_cons = sn.add_constant(X)

In [25]:
logit = sm.Logit(y,X_cons).fit()

Optimization terminated successfully.
         Current function value: 0.264310
         Iterations 8


In [26]:
logit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:           Role_Manager   No. Observations:                 1589
Model:                          Logit   Df Residuals:                     1581
Method:                           MLE   Df Model:                            7
Date:                Sat, 27 Jan 2024   Pseudo R-squ.:                  0.4809
Time:                        21:14:55   Log-Likelihood:                -419.99
converged:                       True   LL-Null:                       -809.08
Covariance Type:            nonrobust   LLR p-value:                9.494e-164
=======================================================================================
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  -7.9272      0.939     -8.443      0.000      -9.767      -6.087
College                 0.0988      0.107      0.922      0.357      -0.111       0.309
City                   -0.3172      0.180     -1.760      0.078      -0.671       0.036
Previous CTC           -0.0001   1.61e-05     -8.350      0.000      -0.000      -0.000
Previous job change    -0.0429      0.080     -0.539      0.590      -0.199       0.113
Graduation Marks       -0.0022      0.006     -0.366      0.714      -0.014       0.010
EXP (Month)            -0.0550      0.007     -7.990      0.000      -0.068      -0.042
CTC                     0.0002   1.13e-05     18.139      0.000       0.000       0.000
=======================================================================================
"""

# Predicting and confusion matrix

In [27]:
clf_lrs.predict_proba(X)

array([[0.89564409, 0.10435591],
       [0.97460934, 0.02539066],
       [0.90364341, 0.09635659],
       ...,
       [0.46810881, 0.53189119],
       [0.8333207 , 0.1666793 ],
       [0.55531659, 0.44468341]])

In [28]:
y_pred = clf_lrs.predict(X)

In [29]:
y_pred

array([0, 0, 0, ..., 1, 0, 0], dtype=int64)

In [30]:
y_pred_03 = (clf_lrs.predict_proba(X)[:,1] >= 0.3)

In [31]:
y_pred_03

array([False, False, False, ...,  True, False,  True])

In [32]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y, y_pred)

array([[1192,   69],
       [ 130,  198]], dtype=int64)

In [33]:
confusion_matrix(y, y_pred_03)

array([[1108,  153],
       [  73,  255]], dtype=int64)

# ACCURACY OF LR MODEL

In [34]:
from sklearn.metrics import accuracy_score, confusion_matrix

In [35]:
accuracy_score(y, y_pred_03)

0.8577721837633732

In [36]:
accuracy_score(y, y_pred)

0.8747640025173065

# Performance metrics
* Precision
$$\textrm{precision} = \frac{\textrm{True Positives}}{\textrm{True Positives} + \textrm{False Positives}}$$
* Recall
$$\textrm{recall} = \frac{\textrm{True Positives}}{\textrm{True Positives} + \textrm{False Negatives}}$$
* AUC (ROC) 

In [37]:
from sklearn.metrics import precision_score, recall_score
from sklearn.metrics import roc_auc_score

In [38]:
precision_score(y, y_pred)

0.7415730337078652

In [39]:
recall_score(y, y_pred)

0.6036585365853658

In [40]:
roc_auc_score(y, y_pred)

0.774470029593238

# Linear Discriminant Ananlysis (LDA)

In [41]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score, confusion_matrix

In [42]:
clf_lda = LinearDiscriminantAnalysis()

In [43]:
clf_lda.fit(X, y)

LinearDiscriminantAnalysis()

In [44]:
y_pred_lda = clf_lda.predict(X)

In [45]:
y_pred_lda

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [46]:
confusion_matrix(y, y_pred_lda)

array([[1209,   52],
       [ 131,  197]], dtype=int64)

# ACCURACY OF LDA MODEL

In [47]:
accuracy_score(y, y_pred_lda)

0.8848332284455632

# Traning our data based on above

In [48]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [49]:
print (X_train.shape,X_test.shape,y_train.shape, y_test.shape)

(1271, 7) (318, 7) (1271,) (318,)


In [50]:
clf_LR = LogisticRegression()
clf_LR.fit(X_train,y_train)

LogisticRegression()

In [51]:
y_test_pred = clf_LR.predict(X_test)

In [52]:
y_test_pred

array([0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [53]:
confusion_matrix(y_test, y_test_pred)

array([[241,  13],
       [ 24,  40]], dtype=int64)

# ACCURACY OF LDA MODEL

In [54]:
from sklearn.metrics import accuracy_score, confusion_matrix
accuracy_score(y_test, y_test_pred)

0.8836477987421384

# K nearest neighbour (KNN)

In [55]:
from sklearn import preprocessing
scaler = preprocessing.StandardScaler().fit(X_train)
X_train_s= scaler.transform(X_train)

In [57]:
scaler = preprocessing.StandardScaler().fit(X_test)
X_test_s= scaler.transform(X_test)

In [58]:
X_test_s

array([[-1.11105684, -1.05830052, -1.93930664, ...,  1.57030415,
         0.7465969 , -0.65242671],
       [ 1.29245387,  0.94491118, -0.73260555, ..., -0.0403075 ,
        -1.47628205, -0.24277797],
       [ 0.09069852,  0.94491118, -1.8656114 , ...,  0.16101896,
        -1.04604742,  0.00824831],
       ...,
       [-1.11105684, -1.05830052,  0.3600471 , ..., -0.37585159,
         0.17295072,  0.32648902],
       [ 1.29245387,  0.94491118,  1.83473087, ...,  0.56367187,
        -1.47628205, -1.84008922],
       [-1.11105684,  0.94491118, -0.33561725, ..., -1.31537505,
         1.46365462, -0.71100003]])

In [59]:
from sklearn.neighbors import KNeighborsClassifier

clf_knn_1 = KNeighborsClassifier(n_neighbors=1)
clf_knn_1.fit(X_train_s, y_train)

KNeighborsClassifier(n_neighbors=1)

In [60]:
confusion_matrix(y_test, clf_knn_1.predict(X_test_s))

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


array([[240,  14],
       [ 22,  42]], dtype=int64)

# ACCURACY OF K-Nearest neighbour (KNN) MODEL

In [61]:
accuracy_score(y_test, clf_knn_1.predict(X_test_s))

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


0.8867924528301887

In [62]:
clf_knn_3 = KNeighborsClassifier(n_neighbors=3)
clf_knn_3.fit(X_train_s, y_train)
accuracy_score(y_test, clf_knn_3.predict(X_test_s))

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


0.8962264150943396

In [63]:
# Model for multiple values of K

from sklearn.model_selection import GridSearchCV
params = {'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20,30]}

In [64]:
grid_search_cv = GridSearchCV(KNeighborsClassifier(), params)
grid_search_cv.fit(X_train_s, y_train)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdim

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdim

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdim

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdim

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20,
                                         30]})

In [65]:
grid_search_cv.best_params_

{'n_neighbors': 7}

In [66]:
optimised_KNN = grid_search_cv.best_estimator_

In [67]:
y_test_pred = optimised_KNN.predict(X_test_s) 

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


In [68]:
confusion_matrix(y_test, y_test_pred)

array([[249,   5],
       [ 29,  35]], dtype=int64)

In [69]:
accuracy_score(y_test, y_test_pred)

0.8930817610062893

# Logistic Regression (LR)   performance                             : 0.8747
# Linear Discriminant Ananlysis (LDA) performance           : 0.8836
# K-Nearest neighbour (KNN) performance                           : 0.8930

# Hence K-Nearest neighbour (KNN) is the best